### PR TITLE
Add Files Tab views

### PR DESCRIPTION
**Pull request by vanwagonet**
refs: MBL-11321
affects: none
release note: none

### CODE REVIEW

**Comment by armstrongnate** [comment]
Did you want to include UI tests with this?

**Comment by armstrongnate** [comment]
I think you had also mentioned adding some presenter tests. We waiting on those?

**Comment by vanwagonet** [comment]
yeah, I'll want to rebase this after #45, and add tests.

File: Student/Student/Submissions/SubmissionDetails/SubmissionDetailsPresenter.swift [hunk 1]
```diff
@@ -64,22 +70,54 @@ class SubmissionDetailsPresenter {
...
+        let assignment = self.assignment.first
+        let submission = submissions.filter({ $0.attempt == selectedAttempt }).first ?? submissions.first
+        selectedAttempt = submission?.attempt ?? selectedAttempt
+        if submission?.attachments?.contains(where: { $0.id == selectedFileID }) != true {
+            selectedFileID = submission?.attachments?.sorted(by: { $0.id < $1.id }).first?.id
+        }
+
+        let assignmentChanged = (assignment != currentAssignment)
+        let fileIDChanged = (currentFileID != selectedFileID)
+        let submissionChanged = (submission != currentSubmission)
```
Comment (armstrongnate) [inline Student/Student/Submissions/SubmissionDetails/SubmissionDetailsPresenter.swift]: Nitpick: Parens on these 3 lines are not needed. I do appreciate them being split up.

File: Student/Student/Submissions/SubmissionDetails/SubmissionDetailsPresenter.swift [hunk 2]
```diff
@@ -64,22 +70,54 @@ class SubmissionDetailsPresenter {
...
+        currentFileID = selectedFileID
+        currentSubmission = submission
+
+        if submissionChanged {
+            view?.embedInDrawer(viewControllerForDrawer())
         }
-        if attempt == 0 {
-            return submissions.first
+        if assignmentChanged || fileIDChanged || submissionChanged {
+            view?.embed(viewControllerForContent())
```
Comment (armstrongnate) [inline Student/Student/Submissions/SubmissionDetails/SubmissionDetailsPresenter.swift]: I like that you're checking for changes before embedding.

File: Student/StudentUITests/Submissions/SubmissionDetails/SubmissionDetailsElement.swift [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+    static var drawerCommentsButton: Element {
+        return ChildElement(parentID: SubmissionDetailsElement.drawerSegmentedControl.id, segmentAt: 0)
+    }
+    static var drawerFilesButton: Element {
+        return ChildElement(parentID: SubmissionDetailsElement.drawerSegmentedControl.id, segmentAt: 1)
+    }
+    static var drawerRubricButton: Element {
+        return ChildElement(parentID: SubmissionDetailsElement.drawerSegmentedControl.id, segmentAt: 2)
+    }
```
Comment (armstrongnate) [inline Student/StudentUITests/Submissions/SubmissionDetails/SubmissionDetailsElement.swift]: These are nice.

File: Student/StudentUITests/Submissions/SubmissionFiles/SubmissionFilesElement.swift [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+import Foundation
+
+class SubmissionFilesElement: RawRepresentable, Element {
```
Comment (armstrongnate) [inline Student/StudentUITests/Submissions/SubmissionFiles/SubmissionFilesElement.swift]: I like this pattern.

File: Student/StudentUITests/Submissions/SubmissionFiles/SubmissionFilesTests.swift [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+import XCTest
+@testable import Core
+import TestsFoundation
+
+class SubmissionFilesTests: StudentTest {
+    lazy var course: APICourse = seedClient.createCourse()
+    lazy var student: AuthUser = createStudent(in: course)
+    lazy var assignment: APIAssignment = seedClient.createAssignment(for: course, submissionTypes: [ .online_upload ], allowedExtensions: [ "pdf" ])
+
+    func testFilesList() {
```
Comment (armstrongnate) [inline Student/StudentUITests/Submissions/SubmissionFiles/SubmissionFilesTests.swift]: Great test! 👏

File: Student/StudentUnitTests/Submissions/SubmissionDetails/SubmissionDetailsPresenterTests.swift [hunk 6]
```diff
@@ -20,29 +20,34 @@ import Core
...
+        embeddedInDrawer = controller
+    }
+
+    var didReload = false
     func reload() {
         didReload = true
-        reloaded.fulfill()
     }
 
+    var didReloadNavbar = false
```
Comment (armstrongnate) [inline Student/StudentUnitTests/Submissions/SubmissionDetails/SubmissionDetailsPresenterTests.swift]: I don't know why but I, too, like the properties right next to the functions like this.

File: Student/Student/Submissions/SubmissionDetails/SubmissionDetailsPresenter.swift (hunk 1 above)
Comment (vanwagonet) [reply Student/Student/Submissions/SubmissionDetails/SubmissionDetailsPresenter.swift]: yeah I debated that, but used them to make it obvious the `!=` has higher precedence than `=`.

**Comment by inst-danger** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>

  One or more files in Core are below the minimum test coverage 50%
  </td>
    </tr>
  </tbody>
</table>


<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>

  This pull request will not generate a release note.
  </td>
    </tr>
  </tbody>
</table>


#### Affected Apps: None

[MBL-11321](https://instructure.atlassian.net/browse/MBL-11321)

Coverage | New % | Master % | Delta
-------- | ----- | -------- | -----
**Core** | 92.57% | 92.81% | -0.24%
Core/Core/Database/CoreData/Models/File.swift | 41.89% | 51.67% | -9.77%
Core/Core/Extensions/OperationQueueExtensions.swift | 0% | 0% | 0%
**Student** | 94.68% | 94.1% | 0.58%
**Teacher** | 100% | 100% | 0%
**React Native** | 92.54% | 92.54% | 0%
<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

**Pull request merged**